src/session.rs

```diff
@@ -166,31 +180,27 @@ impl BackgroundSession {
     ) -> io::Result<BackgroundSession> {
         let mountpoint = se.mountpoint().to_path_buf();
         // Take the fuse_session, so that we can unmount it
-        let fuse_session = se.ch.fuse_session;
-        let fd = se.ch.fd;
-        se.ch.fuse_session = ptr::null_mut();
+        let mount = std::mem::take(&mut se.mount);
+        let mount = mount.ok_or_else(|| io::Error::from_raw_os_error(libc::ENODEV))?;
         let guard = thread::spawn(move || {
             let mut se = se;
             se.run()
         });
         Ok(BackgroundSession {
             mountpoint,
             guard,
-            fuse_session,
-            fd,
+            _mount: mount,
         })
     }
-}
-
-impl Drop for BackgroundSession {
-    fn drop(&mut self) {
-        info!("Unmounting {}", self.mountpoint.display());
-        // Unmounting the filesystem will eventually end the session loop,
-        // drop the session and hence end the background thread.
-        match channel::unmount(&self.mountpoint, self.fuse_session, self.fd) {
-            Ok(()) => (),
-            Err(err) => error!("Failed to unmount {}: {}", self.mountpoint.display(), err),
-        }
+    /// Unmount the filesystem and join the background thread.
+    pub fn join(self) {
+        let Self {
+            mountpoint: _,
+            guard,
+            _mount,
+        } = self;
+        drop(_mount);
+        guard.join().unwrap().unwrap();
```

Comment:
It looks like users need to call `join()` instead of just dropping the `BackgroundSession` now, which will break existing code. Why change it?

Reply:
The previous `Drop` impl doesn't call join AFAICS.  I think the previous behaviour was to attempt the unmount on drop.  This change preserves this behaviour.

The `join` method differs from drop in that unlike before we *do* wait for the background thread to complete. Unless I'm misunderstanding?

Reply:
Oh yes, my bad, I didn't think this through and thought that without `impl Drop for BackgroundSession` it wouldn't unmount. However, `Drop` is implemented on the `Mount` object now